ttnn/cpp/ttnn/tensor/types.cpp

```diff
@@ -108,28 +108,41 @@ bool operator!=(const Padding& padding_a, const Padding& padding_b) { return not
 Shape::Shape(const std::initializer_list<uint32_t> dimensions) :
     rank_(dimensions.size()), dimensions_{}, padding_(dimensions.size()) {
     std::copy(std::begin(dimensions), std::end(dimensions), std::begin(this->dimensions_));
+    validate();
 }
 Shape::Shape(const std::vector<uint32_t>& dimensions) :
     rank_(dimensions.size()), dimensions_{}, padding_(dimensions.size()) {
     std::copy(std::begin(dimensions), std::end(dimensions), std::begin(this->dimensions_));
+    validate();
 }
 
 Shape::Shape(const std::initializer_list<uint32_t> dimensions, const Padding& padding) :
     rank_(dimensions.size()), dimensions_{}, padding_(padding) {
     TT_ASSERT(this->padding_.rank_ == this->rank_);
     std::copy(std::begin(dimensions), std::end(dimensions), std::begin(this->dimensions_));
+    validate();
 }
 Shape::Shape(const std::vector<uint32_t>& dimensions, const Padding& padding) :
     rank_(dimensions.size()), dimensions_{}, padding_(padding) {
     TT_ASSERT(this->padding_.rank_ == this->rank_);
     std::copy(std::begin(dimensions), std::end(dimensions), std::begin(this->dimensions_));
+    validate();
 }
 
 Shape::Shape(const Shape& other, const Padding& padding) :
     dimensions_(other.dimensions_), rank_(other.rank_), padding_(padding) {
     TT_ASSERT(this->padding_.rank_ == this->rank_);
 }
 
+void Shape::validate() const {
+    TT_FATAL(this->rank_ > 0, "Shape rank == 0 is not supported, {}", *this);
+    for (auto index = 0; index < this->rank_; index++) {
+        TT_FATAL(
+            this->dimensions_[index] >= 0,
+            fmt::format("Shape can't have negative dimensions, however dimension {} is {}, {}", index, this->dimensions_[index], *this));
+    }
```

Comment:
Does this mean we'll never implement operators whose output shapes can have 0 size dimensions? [`argwhere`](https://pytorch.org/docs/stable/generated/torch.argwhere.html#torch.argwhere) comes to mind, and other similar operations whose output dimension sizes depend on how many values satisfy certain criteria.

Reply:
Actually, scratch that question. I didn't notice the `>= 0`, I just read in the problem description where it says

> We do not support 0 dimensions

I think that needs to be removed since it appears that we do support them.

Reply:
Also want to point out that `this->dimensions_[index] >= 0` will never be false, since the return type is `uint32_t`, so this expensive loop doesn't have any value (It's expensive because `operator[]` calls `without_padding` which allocates a `std::vector<std::uint32_t>` to construct a new `Shape` every time)

Reply:
Oh! Thank you for spotting! My bet. It actually has to be > 0.
We have a code in the codebase which relying on the fact that its not 0.

Reply:
I had my wires crossed. The `ttnn::types::Shape::operator[]` is the expensive one, not `tt::tt_metal::Shape::operator[]`. I think the `without_padding` method needs to be fixed but that can be done in another PR.